### PR TITLE
[PUBDEV-4676] Allow to handle sparse vectors as the input of the exte…

### PR DESCRIPTION
**Pull request by jakubhava**
…rnal frame handler

### CODE REVIEW

**Comment by jakubhava** [comment]
@h2o-ops please test!

**Comment by jakubhava** [comment]
@mmalohlava this is ready for the review. ( Though there is something failing on sparkling water side, so I will wait for merging once everything is passing)

**Comment by jakubhava** [comment]
@mmalohlava so problems resolved, this is a final version

**Comment by jakubhava** [comment]
@h2o-ops please test!

**Comment by jakubhava** [comment]
All tests passing on this branch

**Comment by jakubhava** [comment]
Thanks @mmalohlava !

File: h2o-core/src/main/java/water/ExternalFrameUtils.java [hunk 1]
```diff
@@ -46,57 +52,87 @@ public static ByteChannel getConnection(String ipPort) throws IOException{
...
+                    break;
+                case EXPECTED_TIMESTAMP:
+                    vecTypes[currentVecIdx] = Vec.T_TIME;
+                    currentVecIdx++;
+                    break;
+                case EXPECTED_VECTOR:
+                    for (int j = 0; j < vecElemSizes[vectorCount]; j++) {
+                        vecTypes[currentVecIdx] = Vec.T_NUM;
+                        currentVecIdx++;
+                        break;
```
Comment (mmalohlava) [inline h2o-core/src/main/java/water/ExternalFrameUtils.java]: This is problem! We really need lot of tests for this stuff to avoid this kind of mistakes,

**Comment by mmalohlava** [comment]
I am worried that this PR was not complete.

File: h2o-core/src/main/java/water/ExternalFrameUtils.java [hunk 2]
```diff
@@ -46,57 +52,87 @@ public static ByteChannel getConnection(String ipPort) throws IOException{
...
-                case EXPECTED_INT: vecTypes[i] = Vec.T_NUM; break;
-                case EXPECTED_LONG: vecTypes[i] = Vec.T_NUM; break;
-                case EXPECTED_FLOAT: vecTypes[i] = Vec.T_NUM; break;
-                case EXPECTED_DOUBLE: vecTypes[i] = Vec.T_NUM; break;
-                case EXPECTED_STRING: vecTypes[i] = Vec.T_STR; break;
-                case EXPECTED_TIMESTAMP: vecTypes[i] = Vec.T_TIME; break;
-                default: throw new IllegalArgumentException("Unknown expected type: "+expectedTypes[i]);
+    public static byte[] vecTypesFromExpectedTypes(byte[] expectedTypes, int[] vecElemSizes){
+        int size = expectedTypes.length;
+        if(vecElemSizes != null && vecElemSizes.length == 0){
```
Comment (mmalohlava) [inline h2o-core/src/main/java/water/ExternalFrameUtils.java]: This is wrong as well, the if condition should be valid only if vecElemSizes.length > 0

Furthermore, it should pass an empty array.

File: h2o-core/src/main/java/water/ExternalFrameUtils.java (hunk 1 above)
Comment (jakubhava) [inline h2o-core/src/main/java/water/ExternalFrameUtils.java]: Yes, bug, should not contain break and should increase vectorCount

File: h2o-core/src/main/java/water/ExternalFrameUtils.java (hunk 2 above)
Comment (jakubhava) [inline h2o-core/src/main/java/water/ExternalFrameUtils.java]: Yes, good catch. I wanted to allow null array to demonstrate there are no vecs but empty array would do it as well. We can remove the nullability in next iteration

**Comment by jakubhava** [comment]
@mmalohlava thanks for pointing out the bugs. It's definitely my bad for introducing PR with these bugs, will ensure to write more tests next time and when I'm back.

However, this PR was also merged without the reviews ( I claimed that's this PR is fine but I was wrong, so again my bad ). I'm not saying it's not my bad, but I think that the responsibility after merging should not be only the authors...

**Comment by mmalohlava** [comment]
It is ok, Kuba - we just need to better test all these new features. I reviewed the code and did not see any problems, i should test it with SW directly.

**Comment by jakubhava** [comment]
Oki, I will ensure to test my code properly for new futures and backwards as well. It's true that I haven't introduced any new tests which is bad. Will do it as the first thing when I return. It's true that the current set of tests is not enough.

Thanks for reviews, I didn't know you already did the review before, it felt like this was first code check :) sorry about that

**Comment by mmalohlava** [comment]
Kuba, no problem, development is iterative process and bugs are common, we will improve our process step-by-step by more testing, more measurements. We are on the good way.

**Pull request merged**